core/src/service/sync_service.rs

```diff
@@ -24,122 +28,228 @@ pub struct WorkCalculated {
     pub most_recent_update_from_server: u64,
 }
 
+#[derive(Clone)]
 pub struct SyncProgress {
     pub total: usize,
     pub progress: usize,
     pub current_work_unit: ClientWorkUnit,
 }
 
-fn should_pull_document(
-    maybe_base_hmac: Option<&DocumentHmac>, maybe_remote_hmac: Option<&DocumentHmac>,
-) -> bool {
-    match (maybe_base_hmac, maybe_remote_hmac) {
-        (_, None) => false,
-        (None, _) => true,
-        (Some(base_hmac), Some(remote_hmac)) => base_hmac != remote_hmac,
-    }
+enum SyncOperation {
+    PullMetadataStart,
+    PullMetadataEnd(Vec<File>),
+    PushMetadataStart(Vec<File>),
+    PushMetadataEnd,
+    PullDocumentStart(File),
+    PullDocumentEnd,
+    PushDocumentStart(File),
+    PushDocumentEnd,
 }
 
-enum SyncProgressOperation {
-    IncrementTotalWork(usize),
-    StartWorkUnit(ClientWorkUnit),
+fn get_work_units(op: &SyncOperation) -> Vec<WorkUnit> {
+    let mut work_units: Vec<WorkUnit> = Vec::new();
+    match op {
+        SyncOperation::PullMetadataEnd(files) => {
+            for file in files {
+                work_units.push(WorkUnit::ServerChange { metadata: file.clone() });
+            }
+        }
+        SyncOperation::PushMetadataStart(files) => {
+            for file in files {
+                work_units.push(WorkUnit::LocalChange { metadata: file.clone() });
+            }
+        }
+        SyncOperation::PullMetadataStart
+        | SyncOperation::PushMetadataEnd
+        | SyncOperation::PullDocumentStart(_)
+        | SyncOperation::PullDocumentEnd
+        | SyncOperation::PushDocumentStart(_)
+        | SyncOperation::PushDocumentEnd => {}
+    }
+    work_units
 }
```

Comment:
sync reports its SyncOperations; some of these become work units (for work calculated) and some become events for reporting sync progress